setup.py

```diff
@@ -0,0 +1,227 @@
+WarpX_QED = os.environ.get('WarpX_QED', 'OFF')
+# TODO: QED tables (require boost)
+WarpX_DIMS = os.environ.get('WarpX_DIMS', '2;3;RZ')
+BUILD_PARALLEL = os.environ.get('BUILD_PARALLEL', '2')
+BUILD_SHARED_LIBS = os.environ.get('WarpX_BUILD_SHARED_LIBS',
+                                   'OFF')
+#BUILD_TESTING = os.environ.get('WarpX_BUILD_TESTING',
+#                               'OFF')
+#BUILD_EXAMPLES = os.environ.get('WarpX_BUILD_EXAMPLES',
+#                                'OFF')
+# openPMD-api sub-control
+HDF5_USE_STATIC_LIBRARIES = os.environ.get('HDF5_USE_STATIC_LIBRARIES', 'OFF')
+ADIOS_USE_STATIC_LIBS = os.environ.get('ADIOS_USE_STATIC_LIBS', 'OFF')
+
+# https://cmake.org/cmake/help/v3.0/command/if.html
+if WarpX_MPI.upper() in ['1', 'ON', 'TRUE', 'YES']:
+    WarpX_MPI = "ON"
+else:
+    WarpX_MPI = "OFF"
+
+cxx_modules = []  # allowed values: warpx_2d, warpx_3d, warpx_rz
+for dim in [x.lower() for x in WarpX_DIMS.split(';')]:
+    name = dim if dim == "rz" else dim + "d"
+    cxx_modules.append(CMakeExtension("warpx_" + name))
+
+# Get the package requirements from the requirements.txt file
+install_requires = []
+with open('./requirements.txt') as f:
+    install_requires = [line.strip('\n') for line in f.readlines()]
+    if WarpX_MPI == "ON":
+        install_requires.append('mpi4py>=2.1.0')
+
+# keyword reference:
+#   https://packaging.python.org/guides/distributing-packages-using-setuptools
+setup(
+    name='pywarpx',
+    # note PEP-440 syntax: x.y.zaN but x.y.z.devN
+    version = '21.01',
+    packages = ['pywarpx'],
+    package_dir = {'pywarpx': 'Python/pywarpx'},
+    author='Jean-Luc Vay, David P. Grote, Maxence Thévenet, Rémi Lehe, Andrew Myers, Weiqun Zhang, Axel Huebl, et al.',
+    author_email='jlvay@lbl.gov, grote1@llnl.gov, maxence.thevenet@desy.de, rlehe@lbl.gov, atmyers@lbl.gov, WeiqunZhang@lbl.gov, axelhuebl@lbl.gov',
+    maintainer='Axel Huebl, David P. Grote, Rémi Lehe', # wheel/pypi packages
+    maintainer_email='axelhuebl@lbl.gov, grote1@llnl.gov, rlehe@lbl.gov',
+    description='WarpX is an advanced electromagnetic Particle-In-Cell code.',
+    long_description=long_description,
+    long_description_content_type='text/markdown',
+    keywords=('WarpX openscience mpi hpc research pic particle-in-cell '
+              'plasma laser-plasma accelerator modeling simulation'),
+    url='https://ecp-warpx.github.io',
+    project_urls={
+        'Documentation': 'https://warpx.readthedocs.io',
+        'Doxygen': 'https://warpx.readthedocs.io/en/latest/_static/doxyhtml/index.html',
+        #'Reference': 'https://doi.org/...', (Paper and/or Zenodo)
+        'Source': 'https://github.com/ECP-WarpX/WarpX',
+        'Tracker': 'https://github.com/ECP-WarpX/WarpX/issues',
+    },
+    ext_modules=cxx_modules,
+    cmdclass=dict(build_ext=CMakeBuild),
+    # scripts=['warpx_2d', 'warpx_3d', 'warpx_rz'],
+    zip_safe=False,
+    python_requires='>=3.6, <3.10',
```

Comment:
Why specify '<3.10'?

Reply:
I usually do this because I want to test a new python release and then push a new package release that supports it first.

It's probably not needed here though (yet), but I know that e.g. middleware like pybind11 needs to be tested against a new CPython usually first.

Reply:
I thought that you might have had a specific reason to exclude 3.10 (like you knew about some upcoming API change that would break things).

I would be less conservative (and leave out the <3.10) but don't have a strong preference. Either way you'll get complaints from users, that 3.10 causes problems, or that 3.10 is not supported (even though it might work fine).

Reply:
From experience, I know that CPython 3.8 had a performance change in memory buffers and CPython 3.9.0 had a memory error on macOS. Both of them were with pybind11 visible through openPMD-api, so I just kept the workflow :)

> Either way you'll get complaints from users, that 3.10 causes problems, or that 3.10 is not supported (even though it might work fine).

Just some subjective thoughts here on my experience with openPMD-api shippings:

True, but users that use conda-forge, brew, modules or their system python are "slow". For instance, most people even now don't use Python 3.9 yet. If we would be slow in releases (monthly is very fast), package managers like conda-forge would just automatically downgrade their python version and users might not even notice if they build a new WarpX environment that it's not cutting edge. The users that go very cutting edge on Python releases are also usually in the category that know how to downgrade and will not be blocked. The broad majority that just gets a breaking default install because a new breaking dependency was released on Saturday evening might be blocked by those - and also adds an urgency to bug reports that is harder to schedule for us maintainers (because they are blocked and we need to move faster) :)